dsc/base.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, final
+
+from dsc.item_submission import ItemSubmission
+from dsc.utilities.aws.s3 import S3Client
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+logger = logging.getLogger(__name__)
+
+
+class BaseWorkflow(ABC):
+    """A base workflow class from which other workflow classes are derived."""
+
+    def __init__(
+        self,
+        email_recipients: list[str],
+        metadata_mapping: dict,
+        s3_bucket: str,
+        s3_prefix: str | None,
+    ) -> None:
+        """Initialize base instance.
+
+        Args:
+            email_recipients: The email addresses to notify after runs of the workflow.
+            metadata_mapping: A mapping file for generating DSpace metadata from the
+            workflow's source metadata.
+            s3_bucket: The S3 bucket containing bitstream and metadata files for the
+            workflow.
+            s3_prefix: The S3 prefix used for objects in this workflow. This prefix does
+            NOT include the bucket name.
+        """
+        self.email_recipients: list[str] = email_recipients
+        self.metadata_mapping: dict = metadata_mapping
+        self.s3_bucket: str = s3_bucket
+        self.s3_prefix: str | None = s3_prefix
+
+    @final
+    def generate_submission_batch(self) -> Iterator[tuple[str, list[str]]]:
```

Comment:
I'm realizing after my first pass that I have a couple of fundamental questions about this base workflow method.

When this is complete, is it correct that:
1. each `ItemSubmission` instantiated in the loop will have uploaded a JSON file to S3 that is _kind_ of like a SIP for DSS to retrieve and ingest into DSpace
2. it yields the URI of each new `ItemSubmission`, which I believe to be the S3 URI of the JSON file uploaded to S3 for that item
3. it also yields `ItemSubmission.bitstream_uris`,  which are _not_ yet getting set on `ItemSubmission`

My question: what is utilizing the results of this yielding?  Will it be another class method?  or the CLI itself?  Unless I'm grossly misunderstanding this application, wouldn't something need to send these as an SQS message to the DSS input queue?

Related question -- and apologies if this will come later -- what happens with the bistream URIs?  Are they included in the SQS message for DSS?

Reply:
The yielded URIs will be utilized by the `deposit` command to generate the SQS message, I mentioned in the PR description 

> with the expectation that unwritten CLI commands will be the interface with the DSS SQS input/output queues and the SES emails.

The workflow class is intended for just the workflow-opinionated code while the CLI command will be a wrapper with universal functionality (e.g. sending SQS messages, SES emails)

And yes, the bitstream URIs are in the the SQS message as well as the URI of the JSON metadata file that is uploaded to S3, [DSS submission message spec](https://github.com/MITLibraries/dspace-submission-service/blob/main/docs/specifications/submission-message-specification.md) for reference.

The steps would be:
1. `ItemSubmission` instantiated with source metadata and bitstreams URIs
2. DSpace metadata generated from source metadata
3. DSpace metadata uploaded to S3 as JSON file and URI is returned
4. ItemSubmission yields `metadata_uris` and `bitstream_uris to `deposit` CLI command for SQS message
5. `deposit` CLI command generates and send SQS message to DSS

Reply:
Ah, sorry missed that in the PR!  Thanks for pointing that out.

And thanks for sharing the DSS submission message spec.  That document is SUPER helpful, and I think it'd be handy to include in docstrings in _this_ application where applicable.

_And_, thanks for the steps; I had missed that `get_bitstream_uris()` is a method that workflows will need to define, and that's where they come from.

This is quite helpful for helping my clarify my next question: any reason why we couldn't yield the `ItemSubmission` object itself?  If I'm understanding correctly, it will have the `metadata_uri` and the `bitstream_uris`.

Happy to wait on that until the CLI level orchestration is in place, where these would get sent out as SQS messages.  Are you envisioning some kind of `DSSInputMessage` class that could encapsulate preparing the final SQS payload for sending, that would meet the [spec](https://github.com/MITLibraries/dspace-submission-service/blob/main/docs/specifications/submission-message-specification.md)?

Which I think has a natural follow-up question: why couldn't `ItemSubmission` have such a method to perform this work?  Like `to_dss_input_queue_message()`?

Happy to meet and discuss if helpful, or keep hashing out in comments.  When I see a class like `ItemSubmission`, which is responsible for uploading the components of a SIP to S3 (metadata and files), I wonder why it wouldn't have the capabilities of actually preparing (and maybe actually sending?) the SQS message for DSS.

Reply:
> Which I think has a natural follow-up question: why couldn't `ItemSubmission` have such a method to perform this work? Like `to_dss_input_queue_message()`?

I think it certainly could! Let's get @jonavellecuerdo 's thoughts as well but now I'm inclined to build SQS message formatting  into `ItemSubmission` rather than the CLI command

Reply:
I do think yielding the actual `ItemSubmission` instance will allow the object to be accessed outside the context of the `BaseWorkflow.generate_submission_batch` method. 🤔 I also think if all `ItemSubmission` instances must comply with the DSS submission message spec, then--as @ehanson8 writes--it also follows that defining the SQS message formatting in `ItemSubmission` would make sense!